### PR TITLE
Add checkout and order promotions

### PR DESCRIPTION
**Pull request by IKarbowiak**
Allow creating promotions that applies on checkout/order level

Covers https://github.com/saleor/saleor/issues/13330

- The solution allows the creation of promotion rules that define `checkoutAndOrder` discount applied on checkout/order level.
- Current implementation applies the discount only on `Checkout`, it's not applicable on `Order` yet.
- The limit of possible `PromotionRule` with `checkoutAndOrder` predicate was introduced. It's set to 100 for the default.

### Measurement:
In case checkout applies for all 100 rules, the checkout calculation time is extended a bit.

**Case 1**
- checkout with one line
- 100 promotion rules with `checkoutAndOrder` predicate
- all rules applies on the checkout

| Mutation        | Time in s    |
| ------------- | ------------- |
| CheckoutCreate  | 0.638  |
| Checkout delivery method update  | 0.896  |
| Checkout complete  | 0.775  |

**Case 2**
- checkout with 1 line
- no promotion rules

| Mutation        | Time in s    |
| ------------- | ------------- |
| CheckoutCreate  | 0.426  |
| Checkout delivery method update  | 0.569  |
| Checkout complete  | 0.713  |

**Case 3**
- checkout with 100 lines
- 100 promotion rules with `checkoutAndOrder` predicate
- all rules applies on the checkout

| Mutation        | Time in s    |
| ------------- | ------------- |
| Checkout delivery method update  | 0.947  |
| Checkout complete  | 1.64  |

**Case 4**
- checkout with 100 lines
- no promotion rules

| Mutation        | Time in s    |
| ------------- | ------------- |
| Checkout delivery method update  | 0.604  |
| Checkout complete  | 2.2  |

# Impact

- [x] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation: https://github.com/saleor/saleor-docs/pull/1067

# Pull Request Checklist

- [x] Privileged queries and mutations are either absent or guarded by proper permission checks
- [x] Database queries are optimized and the number of queries is constant
- [x] Database migrations are either absent or optimized for zero downtime
- [x] The changes are covered by test cases

### CODE REVIEW

**Review by yellowee — APPROVED** [review]

File: saleor/graphql/discount/types/promotions.py [hunk 1]
```diff
@@ -82,19 +88,34 @@ class PromotionRule(ModelObjectType[models.PromotionRule]):
             AuthorizationFilters.AUTHENTICATED_STAFF_USER,
         ],
     )
+    reward_value = PositiveDecimal(
```
Comment (korycins) [inline saleor/graphql/discount/types/promotions.py:91]: Shoiuldn't we add ADDED_IN + PREVIEW_FEATURE?

Comment (korycins) [reply saleor/graphql/discount/types/promotions.py:91]: Same for `type` and `predicate_type`

File: saleor/graphql/discount/utils.py [hunk 2]
```diff
@@ -28,27 +29,17 @@
 PREDICATE_OPERATOR_DATA_T = list[dict[str, Union[list, dict, str, bool]]]
 
 
+class PredicateObjectType(Enum):
+    CATALOGUE = "catalogue"
+    CHECKOUT = "checkout"
+    ORDER = "order"
```
Comment (korycins) [inline saleor/graphql/discount/utils.py:35]: Do we use all of three? I might be wrong, but I think that we reduced it to two, right?

Comment (korycins) [reply saleor/graphql/discount/utils.py:35]: Also, as we have `predicate` for catalogue, and for order, this enum should operate on `CATALOGUE` and `ORDER`. Is it correct?

**Comment by korycins** [comment]
From the benchmark above:
`add checkout lines with reservations` queries increased from `157` to `181`, where duplicates are increased by `9` queries. It may be worth checking why the number is so high. The rest of the tests increase the queries by 2, and it looks correct

File: saleor/graphql/discount/utils.py (hunk 2 above)
Comment (IKarbowiak) [reply saleor/graphql/discount/utils.py:35]: Those values are used for filtering. For `checkout` we are using `CheckoutDiscountedObjectWhere`, for `order` we will be using `OrderDiscountedObjectWhere`

File: saleor/discount/utils.py [hunk 3]
```diff
@@ -556,6 +729,41 @@ def get_variants_to_promotions_map(
...
+        )
+        .exclude(order_predicate={})
+        .prefetch_related("channels")
+    )
+    rule_to_channel_ids_map = _get_rule_to_channel_ids_map(rules)
+
+    checkout_channel_id = checkout.channel_id
+    currency = checkout.channel.currency_code
+    checkout_qs = Checkout.objects.filter(pk=checkout.pk)
+    for rule in list(rules.iterator()):
```
Comment (korycins) [inline saleor/discount/utils.py:751]: This means, that we will load ALL existing order rules to memory.

File: saleor/checkout/utils.py [hunk 4]
```diff
@@ -743,6 +746,12 @@ def add_voucher_to_checkout(
     checkout_info.voucher = voucher
     checkout_info.voucher_code = voucher_code
 
+    # delete discounts from order promotions as cannot be mixed with vouchers
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/checkout/utils.py:741]: Done

File: saleor/graphql/discount/types/promotions.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/graphql/discount/types/promotions.py:91]: Done

File: saleor/discount/utils.py (hunk 3 above)
Comment (korycins) [reply saleor/discount/utils.py:751]: I've made a test, on 5 mln of checkouts. With `https://github.com/pythonprofilers/memory_profiler`
```
In [5]: from memory_profiler import memory_usage
   ...: from time import sleep
   ...:
   ...: def f():
   ...:     # a function that with growing
   ...:     # memory consumption
   ...:     print(Checkout.objects.count())
   ...:     for a in list(Checkout.objects.iterator()):
   ...:         b = a.pk
   ...:     return None
   ```
   vs 
   ```
   In [5]: from memory_profiler import memory_usage
   ...: from time import sleep
   ...:
   ...: def f():
   ...:     # a function that with growing
   ...:     # memory consumption
   ...:     print(Checkout.objects.count())
   ...:     for a in Checkout.objects.iterator():
   ...:         b = a.pk
   ...:     return None
   ```
   
   `max(memory_usage(f))` returned `3737.390625` MiB for `list(iterator())`, and `235.671875` for `iterator()`
   
   Using `iterator` allows db to stream the data (ref:https://docs.djangoproject.com/en/5.0/ref/models/querysets/#with-server-side-cursors), but having `list(iterator())` will cause that we will load all data into worker memory.

File: saleor/discount/interface.py [hunk 5]
```diff
@@ -88,3 +76,25 @@ def fetch_variant_rules_info(
...
+    promotion: "Promotion", rule: "PromotionRule", translation_language_code: str
+):
+    promotion_translations = [
+        translation
+        for translation in promotion.translations.all()
+        if translation.language_code == translation_language_code
+    ]
+    promotion_translation = (
+        promotion_translations[0] if promotion_translations else None
+    )
```
Comment (kadewu) [inline saleor/discount/interface.py:91]: Maybe something like:
```suggestion
    promotion_translation = promotion.translations.filter(language_code=translation_language_code).first()
```
And same for rule_translation
WDYT?

File: saleor/discount/utils.py [hunk 6]
```diff
@@ -535,6 +721,41 @@ def get_variants_to_promotion_rules_map(
...
+    PromotionRuleChannels = PromotionRule.channels.through.objects.filter(
+        channel_id=checkout_channel_id
+    )
+    rules = (
+        PromotionRule.objects.filter(
+            Exists(promotions.filter(id=OuterRef("promotion_id"))),
+            Exists(PromotionRuleChannels.filter(promotionrule_id=OuterRef("id"))),
+        )
+        .exclude(order_predicate={})
+        .prefetch_related("channels")
```
Comment (kadewu) [inline saleor/discount/utils.py:741]: Is this needed? (I just glanced through the function calls, so I might have missed though)

If it is used, then please note, that iterator is compatible with `prefetch_related` if `chunk_size` is set.

**Review by kadewu — COMMENTED** [review]
Just some possible performance improvments from my site, overall LGTM!

File: saleor/discount/interface.py (hunk 5 above)
Comment (IKarbowiak) [reply saleor/discount/interface.py:91]: `promotion.transaltions` are prefetched, `promotion.translations.filter` will cause extra db query

File: saleor/discount/utils.py (hunk 6 above)
Comment (IKarbowiak) [reply saleor/discount/utils.py]: I removed `prefetch_related` as it's not needed anymore :)

**Review by kadewu — APPROVED** [review]

**Review by tomaszszymanski129 — APPROVED** [review]

**Review by yellowee — APPROVED** [review]

**Pull request merged**